### PR TITLE
[SPARK-33231][SPARK-33262][CORE] Make pod allocation executor timeouts configurable & allow scheduling with pending pods.

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Make pod allocation executor timeouts configurable. Keep all known pods in mind when allocating executors to avoid over-allocating if the pending time is much higher than the allocation interval.

This PR increases the default wait time to 600s from the current 60s.

Since nodes can now remain "pending" for long periods of time, we allow additional batches to be scheduled during pending allocation but keep the total number of pods in account.

### Why are the changes needed?
The current executor timeouts do not match that of all real world clusters especially under load. While this can be worked around by increasing the allocation batch delay, that will decrease the speed at which the total number of executors will be able to be requested.

The increase in default timeout is needed to handle real-world testing environments I've encountered on moderately busy clusters and K8s clusters with their own underlying dynamic scale-up of hardware (e.g. GKE, EKS, etc.)

### Does this PR introduce _any_ user-facing change?

Yes new configuration property

### How was this patch tested?

Updated existing test to use the timeout from the new configuration property. Verified test failed without the update.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -227,6 +227,14 @@ private[spark] object Config extends Logging {
       .checkValue(value => value > 0, "Allocation batch delay must be a positive time value.")
       .createWithDefaultString("1s")
 
+  val KUBERNETES_ALLOCATION_EXECUTOR_TIMEOUT =
+    ConfigBuilder("spark.kubernetes.allocation.executor.timeout")
+      .doc("Time to wait before considering a pending executor timedout.")
+      .version("3.1.0")
+      .timeConf(TimeUnit.MILLISECONDS)
+      .checkValue(value => value > 0, "Allocation batch delay must be a positive time value.")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:235]: Is this `Allocation batch delay`?

**Review by dongjoon-hyun — COMMENTED** [review]
Could you describe the procedure you did? Then, I reproduce your test procedure.
```
### How was this patch tested?

Updated existing test to use the timeout from the new configuration property. Verified test failed without the update.
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 2]
```diff
@@ -227,6 +227,14 @@ private[spark] object Config extends Logging {
...
+      .createWithDefaultString("600s")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:236]: Technically, this is the existing value, `600s`. So, which value did you need to change? I'm curious about your use cases.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 3]
```diff
@@ -214,10 +216,9 @@ private[spark] class ExecutorPodsAllocator(
     }
 
     if (newlyCreatedExecutors.isEmpty
-        && currentPendingExecutors.isEmpty
-        && currentRunningCount < currentTotalExpectedExecutors) {
+        && knownPodCount < currentTotalExpectedExecutors) {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:219]: This is irrelevant to `[SPARK-33231][CORE] Make pod allocation executor timeouts configurable`. Could you make another PR about this logic change?

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:219]: @holdenk . Please split this part.

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:219]: So it's not unrelated, but I can give it a second JIRA? Having more time available means that depending on currentRunningCount can be stale information.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 2 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:236]: I believe the existing value is 60s.

**Comment by holdenk** [comment]
Revert the change in ExecutorPodsAllocatorSuite.scala and then try and run the test

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 1 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:235]: Good catch, thanks.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:236]: Then, we had better mention explicitly that this PR is increasing the value in this PR.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
This PR currently puts three orthogonal themes into one title.
1. Propose a new timeout value (`60s` -> `600s`).
2. Support a new configuration
3. Change the logic without mentioning.
```scala
    if (newlyCreatedExecutors.isEmpty
-         && currentPendingExecutors.isEmpty
-         && currentRunningCount < currentTotalExpectedExecutors) {
+         && knownPodCount < currentTotalExpectedExecutors) {
```

We can revise the PR title and description and keep (1) and (2) in this PR. But, please proceed in a separate JIRA for (3). Thanks.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 2 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:236]: Good call

**Comment by holdenk** [comment]
Added the 2nd JIRA and updated the description @dongjoon-hyun

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 4]
```diff
@@ -214,10 +216,9 @@ private[spark] class ExecutorPodsAllocator(
...
       val numExecutorsToAllocate = math.min(
-        currentTotalExpectedExecutors - currentRunningCount, podAllocationSize)
+        currentTotalExpectedExecutors - knownPodCount, podAllocationSize)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:221]: Could you add a test case for `SPARK-33262`?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @holdenk .
Merged to master.
